src/countBy.test-d.ts

```diff
@@ -0,0 +1,35 @@
+import { countBy } from "./countBy";
+import { pipe } from "./pipe";
+
+describe("dataFirst", () => {
+  test("countBy", () => {
+    const data = ["a", "b", "c", "B", "A", "a"];
+    const result = countBy(data, (x) => x.toLowerCase());
+
+    expectTypeOf(result).toEqualTypeOf<Record<string, number>>();
+  });
+});
```

Comment:
I'm missing tests for when the callback function returns a union literal. Those would return the wrong value right now. See my comment in the function code itself on how to fix it.

```suggestion

  test("literal unions", () => {
    const result = countBy([], (x): x is "cat" | "dog" => "cat");
    expectTypeOf(result).toEqualTypeOf<Partial<Record<"cat" | "dog", number>>>();
  });
});
```

Reply:
Thank you for the test! I believe it needs a small adjustment. The callback in your example is a type guard returning a boolean, but `countBy` expects a callback that returns a `PropertyKey`. I think that it could be revised as follows:

```ts
const data = [] as Array<"cat" | "dog">;
const result = countBy(data, (x) => x);

expectTypeOf(result).toEqualTypeOf<Partial<Record<"cat" | "dog", number>>>();
```